### PR TITLE
noise ROIs : final ROIs review, 3D erosion

### PR DESCRIPTION
**Pull request by benoitberanger**
Hello,

Here is PR that includes some features regarding the PhysIO noiseROIs processing.
There are 3 features :
- Write noiseROIs in a volume, after _threshold_ and _erosion_
- Display theses final ROIs using `spm_orthviews` : original ROI + final ROI as overlay
- Erosion is now 3D, using `spm_erode` (mex file, embedded in SPM)

Also, I would like to (re-)enable the possibility to skip the automatic _reslice_ of the raw noiseROI volume.
But I imagine there is a good reason to disable this possibility.

Benoît

### CODE REVIEW

**Comment by mrikasper** [comment]
Dear Benoît,

thank you very much for your contribution! We are very happy that our tools motivate other people to share their improvements with the community.

Before I can review and merge your changes, could you please provide a commit that adds your name and affiliation to our [Contributor-License-Agreement](https://github.com/translationalneuromodeling/tapas/blob/master/Contributor-License-Agreement.md). I have explained the rationale [here](https://github.com/translationalneuromodeling/tapas/issues/11#issuecomment-435213894).

Thank you and all the best,
Lars

**Comment by benoitberanger** [comment]
Hello Lars,

Thank you for the reminder about the _Contributor-License-Agreement_, I forgot to fill it.
I'm happy to contribute.

Please let me know what you think about the way theses features are implemented.
I just saw there is already a verbosity level in the function `tapas_physio_create_noise_rois_regressors`.
I will use it to enable/disable the display using `spm_orthviews`

Best regards,
Benoît

**Comment by mrikasper** [comment]
Dear Benoît,

very nice code indeed, thank you! The thorough commenting makes it very easy to go through the code.

I would have the following comments and suggestions:

1. Use the `verbose.level` to determine the switch for the ROI plotting commands. If you feel this is a general diagnostics, enable plotting for \>=2, if this is more of a debugging, \>=3 should be in order
2. correct a typo: relice instead of reslice
3. I checked my more verbose version history of the PhysIO repository with respect to the automatic reslicing. 
    - Turns out I always had this line in there, which means the test for geometric equality via the affine matrix has not been tested. 
    - So if you want to remove it, please make sure to test it in advance. 
    - I think my rationale here was that reslicing is anyway necessary, because the masks often stem from the structural scan (e.g. TPMs from unified segmentation) and have different resolutions.
4. I try to populate a function `tapas_physio_review` with plot functionality from within the code that might be useful after having run PhysIO, in order to review certain important diagnostic plots and state whether PhysIO has run properly. 
    - This is useful in particular if one hasn't saved the `.fig` files during execution or works on a cluster without graphics capabilities. 
    - So putting your plot functionality for the noise-roise case in that function would be a nice addition, in order to look at the ROI creation. 
    - If you plan to do that, encapsulating your plot routines into their own functions might avoid code duplication.
    - Also, you would probably have to store the file names of the created image files somewhere in the physio-structure, either in `physio.model.output_noise_rois_masks` or `physio.model.output.noise_rois.masks` or `physio.model.noise_rois.masks` or even `physio.output.model.noise_rois.masks`. I haven't fully thought it through, but `physio.model.output.noise_rois.masks` would probably be my favorite at the moment to disentangle input and output within the physio-structure.

Alright, number 1-3 are important to include, number 4 is optional or an inspiration for future PRs ;-)

All the best,
Lars

File: PhysIO/code/tapas_physio_create_noise_rois_regressors.m [hunk 1]
```diff
@@ -113,18 +113,17 @@
     for iter = 1 : n_voxel_crop(r)
         roi = spm_erode(roi);                    % using spm_erode, a compiled mex file
         % roi= imerode(roi, strel('sphere', 1)); % using imerode (+ strel) from Image Processing Toolbox
-        % NB : the result exactly the same with spm_erode or imerode
+        % NB : the result is exactly the same with spm_erode or imerode
     end
     
-    % Write in a volume noise ROIs, after threshold and erosiob
+    % Write the final noise ROIs in a volume, after relice, threshold and erosion
```
[the remark this answers is not in the document]
Comment (benoitberanger) [reply PhysIO/code/tapas_physio_create_noise_rois_regressors.m]: Thank you !

**Comment by benoitberanger** [comment]
Hello Lars,

Thanks for your support.

1. I will use the `verbose.level`
2. Check the typo : I'll do my best.
3. Reslice ?

This a difficult question. The tissue probability maps do have the same resolution and position as the structural scan.

> hasRoiDifferentGeometry = any(any(abs(Vroi.mat - Vimg(1).mat) > 1e-5)) | ...
> any(Vroi.dim-Vimg(1).dim(1:3));

I'm not sure this test is enough : if your structural scan have been acquired in another MRI session they will have different positions. So you will need a `Coregister : Estimate & Reslice`.

My goal is the user should be able to perform quickly _PhysIO:NoiseROIs_ to test different threshold and erosion. This can be achieved if there is no need for a `Coregistration` job.
What do you think ?
But the way to implement it have to be correctly done with great care.

First, I would like to add some information about what kind of space/resolution is expected for the _Noise ROI image files_, inside the GUI.

Best,
Benoît

**Comment by mrikasper** [comment]
Dear Benoît,

I think point (1) and (2) are solved, great! Concerning (3), the reslicing switch:

> This a difficult question. The tissue probability maps do have the same resolution and position as the structural scan.

Indeed, but for the extraction they will have to match the functional scan, which most likely has a different resolution and often even slice geometry.

> 
> > hasRoiDifferentGeometry = any(any(abs(Vroi.mat - Vimg(1).mat) > 1e-5)) | ...
> > any(Vroi.dim-Vimg(1).dim(1:3));
> 
> I'm not sure this test is enough : if your structural scan have been acquired in another MRI session they will have different positions. So you will need a `Coregister : Estimate & Reslice`.

True! I didn't think of this case.

> 
> My goal is the user should be able to perform quickly _PhysIO:NoiseROIs_ to test different threshold and erosion. This can be achieved if there is no need for a `Coregistration` job.
> What do you think ?

Now I understand the use case. I think the best way to implement it is to have a switch in the GUI for the reslicing, and issue a warning, if the geometries don't match and reslicing is off, but then still try the reslicing as a fallback. I would keep reslicing as the default, and whoever wants to try different preprocessing for the masks as you describe will have to know what they are doing.

> But the way to implement it have to be correctly done with great care.
> 
> First, I would like to add some information about what kind of space/resolution is expected for the _Noise ROI image files_, inside the GUI.

Sounds good. It should be the equivalent geometry (affine trafo and matrix size) to the functional image from which the time series shall be extracted.

All the best,
Lars

**Comment by benoitberanger** [comment]
Hey,

Alright, I will add a switch in the GUI such as : 

Force coregistration ? (defalut=YES)
1. YES : do `Coregister : Estimate & Reslice`
2. NO (you have to know what you're doing): Check geometry
     i. Ok, skip `Coregister` then proceed
     ii. Different geometry, **warning** + do  `Coregister : Estimate & Reslice`

Benoît

**Comment by benoitberanger** [comment]
The switch Yes/No perform Coreg is here : [9f9c322](https://github.com/translationalneuromodeling/tapas/pull/34/commits/9f9c322b51d2e9c7a59d445647cb0b25edd6ea17)

**Comment by benoitberanger** [comment]
Dear Lars,

I replaced :

```
% reslice to same geometry
matlabbatch{1}.spm.spatial.coreg.write.ref = fmri_files(1);
matlabbatch{1}.spm.spatial.coreg.write.source = roi_files(r);
...
```

with :

```
% estimate & reslice to same geometry
matlabbatch{1}.spm.spatial.coreg.estwrite.ref = { sprintf('%s,1',fmri_files{1}) }; % select the first volume
matlabbatch{1}.spm.spatial.coreg.estwrite.source = roi_files(r);
...
```

I checked `spm_run_coreg`, `spm_coreg` and `spm_reslice` functions.
In fact, theses routines will always use the **first** volume in 4D file as "ref".
It means, in the previous version, the reslice is performed on volumes(2:end) + source, with volum(1) as ref, such as : 
```
matlabbatch{1}.spm.spatial.coreg.write.ref = {'4D.nii,1'};
matlabbatch{1}.spm.spatial.coreg.write.source = {
4D.nii,2',
4D.nii,3',
...,
noiseROI.nii,
}
```

I hope I'm clear...

Best,
Benoît

**Comment by benoitberanger** [comment]
Dear Lars,

I think this PR is enough mature to be tested.
Let me know if you need me to add/modify/test anything.

Best,
Benoît

**Comment by mrikasper** [comment]
Dear Benoît,

thank you for the thorough implementation. I merged it now into the [development branch](https://github.com/translationalneuromodeling/tapas/tree/development) of TAPAS, will add a unit test and then your addition will also become part of the next TAPAS release (and master), presumably around February.

All the best and merry Christmas,
Lars

**Pull request merged**